src/material.rs

```diff
@@ -225,80 +248,46 @@ pub struct Wireframe {
 
 /// Specifies the appearance of a [`Mesh`](struct.Mesh.html).
 #[derive(Clone, Debug, PartialEq)]
-pub enum Material {
-    /// Renders triangle meshes with a solid color or texture.
+pub(crate) enum MaterialType {
     Basic(Basic),
-
-    /// Renders triangle meshes with a custom pipeline with a basic material as
-    /// its input.
     CustomBasic(basic::Custom),
-
-    /// Renders line strip meshes with a solid color and unit width.
     Line(Line),
-
-    /// Renders triangle meshes with the Gouraud illumination model.
     Lambert(Lambert),
-
-    /// Renders triangle meshes with the Phong illumination model.
     Phong(Phong),
-
-    /// Renders triangle meshes with a PBR (physically-based rendering)
-    /// illumination model
     Pbr(Pbr),
-
-    /// Renders [`Sprite`] objects with the given texture.
-    ///
-    /// [`Sprite`]: ../sprite/struct.Sprite.html
     Sprite(Sprite),
-
-    /// Renders the edges of a triangle mesh with a solid color.
     Wireframe(Wireframe),
 }
 
-impl From<Basic> for Material {
-    fn from(params: Basic) -> Material {
-        Material::Basic(params)
-    }
-}
-
-impl From<basic::Custom> for Material {
-    fn from(params: basic::Custom) -> Material {
-        Material::CustomBasic(params)
-    }
-}
-
-impl From<Lambert> for Material {
-    fn from(params: Lambert) -> Material {
-        Material::Lambert(params)
-    }
-}
-
-impl From<Line> for Material {
-    fn from(params: Line) -> Material {
-        Material::Line(params)
-    }
-}
-
-impl From<Phong> for Material {
-    fn from(params: Phong) -> Material {
-        Material::Phong(params)
-    }
+/// Specifies the appearance of a [`Mesh`](struct.Mesh.html).
+#[derive(Debug, Clone, PartialEq)]
+pub struct Material {
+    pub(crate) mat_type: MaterialType,
+    pub(crate) id: usize,
```

Comment:
can we just hash the material instead of introducing the ID?

Reply:
No, we can't. Materials contains `f32`s, that are unhashable. Even if we can find a way around it, our `Texture<[f32; 4]>` is unhashable too because of `f32` again.

Reply:
`f32` in general are unhashable, but only because of the exceptions like de-normalized floats and NaN, Inf, etc. We can safely assume the f32 in the material are not exceptional (and abort otherwise). Hashing can't be derived automatically, but it can be implemented using one of the following methods:
  - just hashing raw 32bits of the float
  - if the float is known to be in a rage (say, 0-1), quantitize it with enough precision (`u32` is more than enough). 

`Texture<[f32; 4]>` is just a handle, once again, that is hashable. Caveat: we need to investigate in depth the safety of hashing the gfx handles, but for now this is an acceptable choice.